### PR TITLE
feat(kuma-cp) coordinates webservice

### LINKED ISSUE #315: `kuma-cp` should provide an endpoint to return locations of individual components, such as `bootstrap server`, `api server`, etc
### Summary

* `kuma-cp` should provide an endpoint to return locations of individual components, such as `bootstrap server`, `api server`, etc 

### Motivation

* enable #314
* allow evolution of Kuma endpoints without a need to modify existing deployments, e.g. migrate `bootstrap server` from `HTTP` to `HTTPS`  

### Example

```
{
  "apis": {
    "bootstrap": {
      "url": "http://kuma-control-plane:5682"
    },
    "initial-token": {
      "url": "https://kuma-control-plane:5683"
    }
  }
}
```

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

Endpoint that serves coordinates of various services in Kuma.

*Questions*
Should Kuma DP also use API Server to retrieve bootstrap coordinates? Right now we configure it with bootstrap url. 

### Issues resolved

Fix #315

### CODE REVIEW

File: pkg/api-server/coordinates_ws.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package api_server
+
+import (
+	kuma_cp "github.com/Kong/kuma/pkg/config/app/kuma-cp"
+	"github.com/Kong/kuma/pkg/coordinates"
+	"github.com/emicklei/go-restful"
+)
+
+func componentsWs(cfg kuma_cp.Config) *restful.WebService {
```
Comment (yskopets) [inline pkg/api-server/coordinates_ws.go]: I don't like the fact that `api-server` now depends on configuration of the entire `Control Plane`.

It makes it seem that `API Server` is inseparable from the `Control Plane`.

I would suggest an alternative solution:
* make configuration of available APIs part of `ApiServerConfig` (effectively, move existing `coordinates.Coordinates` types into `pkg/config/api-server/catalogue`)
* don't map this configuration to YAML or environment variables yet (we reserve this feature for the future but don't want to expose it to the users right away)
* auto-configure  available APIs at `autoconfigure` stage of the `bootstrap` process (effectively, move logic from `pkg/coordinates/types.go` into `pkg/core/bootstrap/autoconfigure.go`)

File: pkg/api-server/coordinates_ws.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	ws := new(restful.WebService).Path("/coordinates")
```
Comment (yskopets) [inline pkg/api-server/coordinates_ws.go]: Maybe `/catalogue` or `/apis/catalogue` ?

File: pkg/config/app/kuma-cp/config.go [hunk 3]
```diff
@@ -74,6 +74,8 @@ type Config struct {
 	Defaults *Defaults `yaml:"defaults"`
 	// Reports configuration
 	Reports *Reports `yaml:"reports"`
+	// Hostname on which Kuma is exposed
+	Hostname string `yaml:"hostname" envconfig:"kuma_hostname"`
```
Comment (yskopets) [inline pkg/config/app/kuma-cp/config.go]: Let's group it under `General` field and move it to the top of `Config`

Comment (yskopets) [reply pkg/config/app/kuma-cp/config.go]: Maybe rename it into one of  `RecommendedHostname`, `AdvertisedHostname`, etc

File: pkg/config/app/kuma-cp/config.go [hunk 4]
```diff
@@ -74,6 +74,8 @@ type Config struct {
 	Defaults *Defaults `yaml:"defaults"`
 	// Reports configuration
 	Reports *Reports `yaml:"reports"`
+	// Hostname on which Kuma is exposed
```
Comment (yskopets) [inline pkg/config/app/kuma-cp/config.go]: Let's change the comment to
```
Hostname that other components should use in order to connect to the Control Plane.

Control Plane will use this value in configuration generated for dataplanes, 
in responses to `kumactl`, etc.   
```

**Comment by yskopets** [comment]
> Should Kuma DP also use API Server to retrieve bootstrap coordinates? Right now we configure it with bootstrap url.

Yes. There is #314 for that

File: pkg/config/app/kuma-cp/config.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/config/app/kuma-cp/config.go]: What goes to General? Should Environment also be in General?

Comment (yskopets) [reply pkg/config/app/kuma-cp/config.go]: Let's start by moving only `AdvertisedHostname` into `General`.

And we can open a separate issue for doing the same for `Environment`.

**Comment by jakubdyszkiewicz** [comment]
moved into general

**Pull request merged**